keps/sig-node/20191002-seccomp-custom-profiles.md

```diff
@@ -0,0 +1,258 @@
+
+
+#### Rollout of profile changes
+
+Seccomp profiles are applied at container creation time, therefore updating an existing user-defined profile will not cause any changes to the running containers that are using it. 
+They will need to be restarted in order for the changes to take effect, which users will have to do manually. 
+
+
+#### Starting containers with non-existent profile
+
+An error message should be returned stating the profile could not be found. Containers targeted for seccomp profiles should not be able to start if their profiles are deleted. 
+
+If a running container has its policy deleted, nothing will happen until it has to be restarted, which in this case would lead to a failure as stated above.
+
+
+
+### User Stories
+
+#### Story 1
+As a user, I want to be able to create and maintain custom seccomp profiles without having to physically copy files across all my existing (and new) nodes.
+
+
+#### Story 2
+As a user, I want to be able to create and maintain custom seccomp profiles by using kubectl, in the same way I do other Kubernetes objects. 
+
+
+#### Story 3
+As a user, I want to be able to determine whether custom seccomp profiles have been changed since applied to running containers. 
+
+
+
+## Design Details
+
+### Test Plan
+
+Seccomp already has E2E tests, but the tests are guarded by the [Feature:Seccomp] tag and not run in the standard test suites.
+
+Prior to [being marked GA](20190717-seccomp-ga.md), the feature tag will be removed from the seccomp tests.
+
+Test coverage for configmap profiles will be added and be tested in isolation. 
+
+E2E tests will be required to ensure profile changes won't impact running containers.
+
+### Upgrade / Downgrade Strategy
+
+No upgrade changes required - both localhost and configmap profiles will be supported in parallel.
+
+### Version Skew Strategy
+
+The new configmap based profiles will only be supported from this version on. Users will need to ensure that pods using the new feature won't be scheduled on nodes running older Kubelet versions.
+
+### Graduation Criteria
+
+- API changes to map seccomp profiles to an additional `Kubernetes` type.
+- E2E tests
+
+
+## Implementation History
+- 2019-10-02: Initial KEP
+- 2019-10-15: Minor changes
+- 2019-10-17: Add alternative to use CRD instead of ConfigMap
+
+
+## Alternatives
+
+**Use CRD instead of ConfigMap.** The decision to use `ConfigMap` was to avoid unnecessary complexity. Below are some key points from the [official guidance](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#should-i-use-a-configmap-or-a-custom-resource) which clarifies the reasoning:
+
+- There is an existing, well-documented config file format.
+- The entire config file will be stored into one key of a configMap.
+- The main use of the config file is for the cluster to consume the file to pass it on to CRIs.
+- Users could potentially perform rolling updates via Deployment, when the file is updated.
+
```

Comment:
Some counter arguments:

- The configmap won't be mounted into a container, so it won't be reusing any Kubelet code.
- A CRD could be cluster-scoped (although I'm on the fence about whether that's actually desirable)
- A configmap can have multiple keys, which might not make sense in this case.
- A dedicated type could be immutable, to avoid some uncertainty around updates.
- A fully-structured CRD could provide static type checking for early failure (although this creates the undesirable tight coupling to the OCI format)

Reply:
Added all but one. I will explain better on the next comment.